arwen/common.go

```diff
@@ -224,3 +227,37 @@ func (ac *AsyncGeneratedCall) GetValueBytes() []byte {
 func (ac *AsyncGeneratedCall) IsInterfaceNil() bool {
 	return ac == nil
 }
+
+func GetStorageLoadCost(trieDepth int64, metering MeteringContext, staticGasCost uint64, enableEpochsHandler vmcommon.EnableEpochsHandler) (uint64, error) {
+	if enableEpochsHandler.IsDynamicGasCostForDataTrieStorageLoadEnabled() {
+		return computeGasForStorageLoadBasedOnTrieDepth(trieDepth, metering.GasSchedule().DynamicStorageLoad)
+	}
+
+	return staticGasCost, nil
+}
+
+func computeGasForStorageLoadBasedOnTrieDepth(trieDepth int64, quadraticCoefficients config.DynamicStorageLoad) (uint64, error) {
+	if trieDepth == 0 {
+		return 0, nil
+	}
+
+	a := getSignedCoefficient(quadraticCoefficients.A, quadraticCoefficients.SignOfA)
+	b := getSignedCoefficient(quadraticCoefficients.B, quadraticCoefficients.SignOfB)
+	c := getSignedCoefficient(quadraticCoefficients.C, quadraticCoefficients.SignOfC)
+
+	fx := a*trieDepth*trieDepth + b*trieDepth + c
+
+	if fx < 0 {
+		return 0, fmt.Errorf("invalid value for gas cost, a = %v, b = %v, c = %v, trie depth = %v", a, b, c, trieDepth)
+	}
+
+	return uint64(fx), nil
```

Comment:
please add protection here + a log.Error print if the result is less than a minimum threshold. In that case you should return the default current storageLoad value. 
Reason: we might forget to define the storage costs for a gas schedule & this function will return 0 which is extremely dangerous. 
The log prints will help us in the debugging/system tests processes

Reply:
Added.